statsd/statsd.go

```diff
@@ -143,37 +143,40 @@ func NewBuffered(addr string, buflen int) (*Client, error) {
 
 // format a message from its name, value, tags and rate.  Also adds global
 // namespace and tags.
-func (c *Client) format(name string, value interface{}, suffix []byte, tags []string, rate float64) string {
-	var buf bytes.Buffer
+func (c *Client) format(name string, value interface{}, suffix []byte, tags []string, rate float64) []byte {
+	// preallocated buffer, stack allocated as long as it doesn't escape
+	buf := make([]byte, 0, 200)
+
 	if c.Namespace != "" {
-		buf.WriteString(c.Namespace)
```

Comment:
Have you tried `strings.Builder`?

Reply:
I didn't try it when originally working on this PR. Go 1.10 (where `strings.Builder` was introduced) had only been out for a month or so at the time.

Theoretically `strings.Builder` should have similar performance, possibly excepting some of the `strconv` calls. If it's more readable using that API my opinion is that it would be worth testing out.

Reply:
I would be curious to hear if `strings.Builder` introduces further improvement. I was surprised to see how big the difference is between it and `bytes.Buffer`. It was supposedly built exactly for this use case.

IMHO given how critical this path is, performance has priority over readability.

Reply:
Having reviewed this code again, I'm doubtful that it would be an improvement but I'd be interested to know if my reasoning is incorrect. Part of the improvement is due to convincing the compiler to stack allocate the buffer by having a fixed initial size. Even with `var b strings.Builder; b.Grow(200)` I'm not sure the compiler would see through it and do something similar.

Ultimately the message is written to the network as a `[]byte`, so I'm not sure there's much to be gained by converting to a string, even with the optimization that `strings.Builder` provides.

Reply:
If the message is processed and used as a `[]byte`, then you're likely right. Although, [recent changes](https://github.com/golang/go/commit/ca0c449a6b1c6ecc75169f93cffa8a5630740030) in the Go compiler indicate the two to be interchangeable.

Honestly, the only way to tell is to implement and run the benchmarks again.